### PR TITLE
Torch lr schedulers

### PR DESCRIPTION
**Pull request by supercgor**
add some learning rate schedulers for torch.

### CODE REVIEW

File: deepxde/model.py [hunk 1]
```diff
@@ -93,7 +93,12 @@ def compile(
                 - For backend PyTorch:
 
                     - `StepLR <https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.StepLR.html>`_: ("step", step_size, gamma)
-
+                    - `CosineAnnealingLR <https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.CosineAnnealingLR.html>`_: ("cosine", T_max, eta_min)
+                    - `InverseTimeLR <https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/schedules/InverseTimeDecay>`_: ("inverse time", decay_steps, decay_rate)
```
Comment (lululxvi) [inline deepxde/model.py:97]: This is TF, not pytorch.

File: deepxde/optimizers/pytorch/optimizers.py [hunk 2]
```diff
@@ -61,9 +61,17 @@ def _get_learningrate_scheduler(optim, decay):
         return None
 
     if decay[0] == "step":
-        return torch.optim.lr_scheduler.StepLR(
-            optim, step_size=decay[1], gamma=decay[2]
-        )
+        return torch.optim.lr_scheduler.StepLR(optim, step_size=decay[1], gamma=decay[2])
```
Comment (lululxvi) [inline deepxde/optimizers/pytorch/optimizers.py:64]: Format via black https://pypi.org/project/black/

File: deepxde/model.py (hunk 1 above)
Comment (supercgor) [reply deepxde/model.py:97]: The reason is PyTorch didn't have a built-in InverseTimeLR. So I just copied a document that may help.

Comment (lululxvi) [reply deepxde/model.py]: OK. Then you can add the link back.

File: deepxde/model.py [hunk 3]
```diff
@@ -93,7 +93,12 @@ def compile(
                 - For backend PyTorch:
 
                     - `StepLR <https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.StepLR.html>`_: ("step", step_size, gamma)
-
+                    - `CosineAnnealingLR <https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.CosineAnnealingLR.html>`_: ("cosine", T_max, eta_min)
+                    - `InverseTimeLR see the document from TensorFlow`_: ("inverse time", decay_steps, decay_rate)
+                    - `ExponentialLR <https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ExponentialLR.html>`_: ("exponential", gamma)
+                    - `LambdaLR <https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.LambdaLR.html>`_: ("lambda", lambda_fn: Callable[[step], float])
+                    - `torch.optim.Optimizer <https://pytorch.org/docs/stable/optim.html#torch.optim.Optimizer>`_: (optimizer, )
```
Comment (lululxvi) [inline deepxde/model.py:100]: Why this torch.optim.Optimizer?

Comment (lululxvi) [reply deepxde/model.py:100]: It is already supported

https://github.com/lululxvi/deepxde/blob/23bba4f983d57eb971759da47a9658d4f6238312/deepxde/optimizers/pytorch/optimizers.py#L15

Comment (supercgor) [reply deepxde/model.py]: I found a mistake there, should be lr scheduler. I've fixed it

File: deepxde/model.py [hunk 4]
```diff
@@ -93,6 +93,11 @@ def compile(
                 - For backend PyTorch:
 
                     - `StepLR <https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.StepLR.html>`_: ("step", step_size, gamma)
+                    - `CosineAnnealingLR <https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.CosineAnnealingLR.html>`_: ("cosine", T_max, eta_min)
+                    - `InverseTimeLR <https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/schedules/InverseTimeDecay>`_: ("inverse time", decay_steps, decay_rate)
+                    - `ExponentialLR <https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ExponentialLR.html>`_: ("exponential", gamma)
+                    - `LambdaLR <https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.LambdaLR.html>`_: ("lambda", lambda_fn: Callable[[step], float])
+                    - `torch.optim.lr_scheduler.LRScheduler <https://pytorch.org/docs/stable/optim.html#torch.optim.Optimizer>`_: (LRScheduler, args, kwargs)
```
Comment (lululxvi) [inline deepxde/model.py:100]: The link is still torch.optim.Optimizer

Comment (supercgor) [reply deepxde/model.py]: I deleted the extra content after the hashtag. LRScheduler doesn't have any documents for the base class. So I used the link of the optim website.

File: deepxde/model.py [hunk 5]
```diff
@@ -93,6 +93,11 @@ def compile(
                 - For backend PyTorch:
 
                     - `StepLR <https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.StepLR.html>`_: ("step", step_size, gamma)
+                    - `CosineAnnealingLR <https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.CosineAnnealingLR.html>`_: ("cosine", T_max, eta_min)
+                    - `InverseTimeLR <https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/schedules/InverseTimeDecay>`_: ("inverse time", decay_steps, decay_rate)
+                    - `ExponentialLR <https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ExponentialLR.html>`_: ("exponential", gamma)
+                    - `LambdaLR <https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.LambdaLR.html>`_: ("lambda", lambda_fn: Callable[[step], float])
+                    - `torch.optim.lr_scheduler.LRScheduler <https://pytorch.org/docs/stable/optim.html>`_: (LRScheduler, args, kwargs)
```
Comment (lululxvi) [inline deepxde/model.py:100]: Is this necessary, as there is already LambdaLR?

File: deepxde/optimizers/pytorch/optimizers.py [hunk 6]
```diff
@@ -64,6 +64,20 @@ def _get_learningrate_scheduler(optim, decay):
         return torch.optim.lr_scheduler.StepLR(
             optim, step_size=decay[1], gamma=decay[2]
         )
+    elif decay[0] == "lambda":
```
Comment (lululxvi) [inline deepxde/optimizers/pytorch/optimizers.py:67]: Reorder different decays to match the order of docs.

File: deepxde/model.py (hunk 5 above)
Comment (supercgor) [reply deepxde/model.py:100]: no, I think lambda can do many cases. shall I just cancel it? Or leave it here.

Comment (lululxvi) [reply deepxde/model.py:100]: Yes, delete it.

File: deepxde/optimizers/pytorch/optimizers.py [hunk 7]
```diff
@@ -64,6 +64,18 @@ def _get_learningrate_scheduler(optim, decay):
         return torch.optim.lr_scheduler.StepLR(
             optim, step_size=decay[1], gamma=decay[2]
         )
+    elif decay[0] == "cosine":
+        return torch.optim.lr_scheduler.CosineAnnealingLR(
+            optim, T_max=decay[1], eta_min=decay[2]
```
Comment (lululxvi) [inline deepxde/optimizers/pytorch/optimizers.py:69]:
```
optim, decay[1], eta_min=decay[2]
```

File: deepxde/optimizers/pytorch/optimizers.py [hunk 8]
```diff
@@ -64,6 +64,18 @@ def _get_learningrate_scheduler(optim, decay):
...
+        )
+    elif decay[0] == "inverse time":
+        return torch.optim.lr_scheduler.LambdaLR(
+            optim, lambda step: 1 / (1 + decay[2] * (step / decay[1]))
+        )
+    elif decay[0] == "exponential":
+        return torch.optim.lr_scheduler.ExponentialLR(optim, gamma=decay[1])
```
Comment (lululxvi) [inline deepxde/optimizers/pytorch/optimizers.py:76]:
```
return torch.optim.lr_scheduler.ExponentialLR(optim, decay[1])
```
